File: api/routes/data_maturity.py
"""API — National Data Maturity Dashboard."""

from __future__ import annotations

from typing import Any

from fastapi import APIRouter
from fastapi.responses import HTMLResponse

from api.services import data_maturity_engine as ndm

router = APIRouter()


@router.get("", summary="Maturité nationale des données")
@router.get("/", summary="Maturité nationale des données")
def get_maturity() -> dict[str, Any]:
    return ndm.build_national_maturity()


@router.get("/details", summary="Détail par référentiel")
def get_details() -> dict[str, Any]:
    return ndm.build_details()


@router.get("/roadmap", summary="Feuille de route data + priorités")
def get_roadmap() -> dict[str, Any]:
    return ndm.build_roadmap_payload()


@router.get("/map", summary="GeoJSON maturité data (≠ couverture radio)")
def get_map() -> dict[str, Any]:
    return ndm.build_map_payload()


@router.get("/report", summary="Rapport Direction (JSON)")
def get_report() -> dict[str, Any]:
    return ndm.build_report_payload()


@router.get("/report.html", summary="Rapport imprimable (HTML → PDF navigateur)", response_class=HTMLResponse)
def get_report_html() -> HTMLResponse:
    report = ndm.build_report_payload()
    score = report.get("national_score")
    band = (report.get("national_band") or {}).get("label") or "—"
    rows = "".join(
        f"<tr><td>{d.get('label')}</td><td>{d.get('score') if d.get('score') is not None else '—'}%</td>"
        f"<td>{(d.get('band') or {}).get('label') or '—'}</td></tr>"
        for d in (report.get("dashboard") or [])
    )
    prios = "".join(
        f"<li>{'★' * int(p.get('stars') or 0)}{'☆' * (5 - int(p.get('stars') or 0))} "
        f"<strong>{p.get('label')}</strong> — {p.get('reason')}</li>"
        for p in (report.get("priorities") or [])
    )
    road = report.get("roadmap") or {}

    def road_block(title: str, items: list) -> str:
        lis = "".join(f"<li><strong>{i.get('action')}</strong> → {i.get('expected_gain')}</li>" for i in (items or []))
        return f"<h3>{title}</h3><ul>{lis or '<li>Aucun item</li>'}</ul>"

    html = f"""<!doctype html>
<html lang="fr"><head><meta charset="utf-8"/>
<title>Rapport National de Maturité des Données</title>
<style>
body{{font-family:Georgia,serif;margin:2rem;color:#0f172a}}
h1{{font-size:1.6rem}} table{{border-collapse:collapse;width:100%;margin:1rem 0}}
th,td{{border:1px solid #cbd5e1;padding:.4rem .6rem;text-align:left}}
th{{background:#f1f5f9}} .muted{{color:#64748b;font-size:.9rem}}
@media print{{button{{display:none}}}}
</style></head><body>
<button onclick="window.print()">Imprimer / PDF</button>
<h1>Rapport National de Maturité des Données</h1>
<p class="muted">SIG-FDSU RDC · Direction · {report.get('_meta', {}).get('generated_at', '')}</p>
<p><strong>Maturité nationale :</strong> {score if score is not None else '—'} % ({band})</p>
<p class="muted">Calcul : moyenne pondérée des scores domaines (dimensions null exclues). Data First.</p>
<h2>Tableau de bord</h2>
<table><thead><tr><th>Référentiel</th><th>Score</th><th>Bande</th></tr></thead>
<tbody>{rows}</tbody></table>
<h2>Données prioritaires à acquérir</h2>
<ul>{prios or '<li>Aucune priorité critique</li>'}</ul>
{road_block('Court terme', road.get('short_term'))}
{road_block('Moyen terme', road.get('medium_term'))}
{road_block('Long terme', road.get('long_term'))}
<p class="muted">{report.get('_meta', {}).get('note') or ''}</p>
</body></html>"""
    return HTMLResponse(content=html)

File: dashboard/modules/shared/data-maturity/data-maturity.js
/**
 * National Data Maturity Dashboard — Salle de Pilotage.
 * Totaux exclusivement depuis /api/data-maturity (Data First).
 */
(function initDataMaturityDashboard(global) {
  const API_BASE = `${global.location.protocol}//${global.location.hostname}:8001`;

  function escapeHtml(value) {
    return String(value ?? '')
      .replace(/&/g, '&amp;')
      .replace(/</g, '&lt;')
      .replace(/>/g, '&gt;')
      .replace(/"/g, '&quot;');
  }

  function fmt(value) {
    if (value == null || value === '') return '—';
    if (typeof value === 'number') return value.toLocaleString('fr-FR');
    return String(value);
  }

  function fetchJson(path) {
    return fetch(`${API_BASE}${path}`, { headers: { Accept: 'application/json' }, cache: 'no-store' })
      .then((r) => {
        if (!r.ok) throw new Error(path);
        return r.json();
      });
  }

  function bandClass(band) {
    const code = (band && band.code) || 'unknown';
    return `ndm-band ndm-band--${escapeHtml(code)}`;
  }

  function stars(n) {
    const s = Math.max(0, Math.min(5, Number(n) || 0));
    return `${'★'.repeat(s)}${'☆'.repeat(5 - s)}`;
  }

  function ensureHost() {
    let host = document.querySelector('#esr-data-maturity');
    if (host) return host;
    const root = document.querySelector('.esr-root .esr-col-main') || document.querySelector('.esr-root');
    if (!root) return null;
    host = document.createElement('section');
    host.id = 'esr-data-maturity';
    host.className = 'esr-card ndm-card';
    host.innerHTML = `
      <header class="esr-card-header">
        <h3>Maturité des Données Nationales</h3>
        <div class="ndm-header-actions">
          <button type="button" class="secondary-button" id="ndm-btn-report">Rapport Direction</button>
          <button type="button" class="secondary-button" id="ndm-btn-refresh">Actualiser</button>
        </div>
      </header>
      <div id="esr-data-maturity-body" class="ndm-body">Chargement…</div>`;
    const maturitySdg = document.querySelector('#esr-sdg-maturity');
    if (maturitySdg && maturitySdg.parentNode) {
      maturitySdg.parentNode.insertBefore(host, maturitySdg);
    } else {
      root.prepend(host);
    }
    return host;
  }

  function renderDetail(domain) {
    if (!domain) return '';
    const dims = Object.values(domain.dimensions || {});
    return `
      <article class="ndm-detail" data-code="${escapeHtml(domain.code)}">
        <header>
          <h4>${escapeHtml(domain.label)} — ${domain.score != null ? `${fmt(domain.score)} %` : '—'}</h4>
          <span class="${bandClass(domain.band)}">${escapeHtml(domain.band?.label || '')}</span>
        </header>
        <p class="ndm-meta">Source : ${escapeHtml(domain.source || '—')} · As-of : ${escapeHtml(domain.as_of || '—')} · Version : ${escapeHtml(domain.version || '—')}</p>
        <p class="ndm-meta">Objets : ${fmt(domain.object_count)} · Relations : ${escapeHtml(domain.relations || '—')}</p>
        <div class="ndm-dims">${dims.map((d) => `
          <span title="${escapeHtml(d.label)}">${escapeHtml(d.label)} <strong>${d.score != null ? fmt(d.score) : '—'}</strong></span>
        `).join('')}</div>
        ${(domain.strengths || []).length ? `<p><em>Forces :</em> ${domain.strengths.map(escapeHtml).join(' · ')}</p>` : ''}
        ${(domain.weaknesses || []).length ? `<p><em>Faiblesses :</em> ${domain.weaknesses.map(escapeHtml).join(' · ')}</p>` : ''}
        ${(domain.anomalies || []).length ? `<p><em>Anomalies :</em> ${domain.anomalies.map(escapeHtml).join(' · ')}</p>` : ''}
        ${(domain.recommendations || []).length ? `<p><em>Recommandations :</em> ${domain.recommendations.map(escapeHtml).join(' · ')}</p>` : ''}
      </article>`;
  }

  function mountDataMaturityDashboard() {
    const host = ensureHost();
    if (!host) return Promise.resolve(null);
    const body = host.querySelector('#esr-data-maturity-body') || host;
    body.innerHTML = '<p class="esr-muted">Calcul de la maturité nationale…</p>';

    return fetchJson('/api/data-maturity')
      .then((payload) => {
        const dash = payload.dashboard || [];
        const priorities = payload.priorities || [];
        const roadmap = payload.roadmap || {};
        body.innerHTML = `
          <div class="ndm-national">
            <div>
              <span class="ndm-national-label">Maturité nationale</span>
              <strong class="ndm-national-score">${payload.national_score != null ? `${fmt(payload.national_score)} %` : '—'}</strong>
              <span class="${bandClass(payload.national_band)}">${escapeHtml(payload.national_band?.label || '')}</span>
            </div>
            <p class="dxl-note">Moyenne pondérée des référentiels scorés — dimensions absentes exclues (Data First).</p>
          </div>
          <div class="ndm-grid">
            ${dash.map((d) => `
              <button type="button" class="ndm-tile ${bandClass(d.band)}" data-ndm-code="${escapeHtml(d.code)}" title="${escapeHtml(d.label)}">
                <span>${escapeHtml(d.label)}</span>
                <strong>${d.score != null ? `${fmt(d.score)} %` : '—'}</strong>
              </button>`).join('')}
          </div>
          <div id="ndm-detail-host" class="ndm-detail-host"></div>
          <h4>Données prioritaires à acquérir</h4>
          <ul class="ndm-priorities">${priorities.map((p) => `
            <li><span class="ndm-stars" aria-label="${p.stars} étoiles">${stars(p.stars)}</span>
              <strong>${escapeHtml(p.label)}</strong> — ${escapeHtml(p.reason || '')}</li>`).join('') || '<li>Aucune priorité critique</li>'}
          </ul>
          <h4>Feuille de route data</h4>
          <div class="ndm-roadmap">
            <div><h5>Court terme</h5><ul>${(roadmap.short_term || []).map((i) => `<li>${escapeHtml(i.action)} <em>(${escapeHtml(i.expected_gain)})</em></li>`).join('') || '<li>—</li>'}</ul></div>
            <div><h5>Moyen terme</h5><ul>${(roadmap.medium_term || []).map((i) => `<li>${escapeHtml(i.action)} <em>(${escapeHtml(i.expected_gain)})</em></li>`).join('') || '<li>—</li>'}</ul></div>
            <div><h5>Long terme</h5><ul>${(roadmap.long_term || []).map((i) => `<li>${escapeHtml(i.action)} <em>(${escapeHtml(i.expected_gain)})</em></li>`).join('') || '<li>—</li>'}</ul></div>
          </div>
          <p class="dxl-note">Calcul : ${escapeHtml(payload._meta?.generated_at || '')}</p>`;

        const detailHost = body.querySelector('#ndm-detail-host');
        const byCode = Object.fromEntries((payload.domains || []).map((d) => [d.code, d]));
        body.querySelectorAll('[data-ndm-code]').forEach((btn) => {
          btn.addEventListener('click', () => {
            const code = btn.getAttribute('data-ndm-code');
            if (detailHost) detailHost.innerHTML = renderDetail(byCode[code]);
          });
        });

        host.querySelector('#ndm-btn-refresh')?.addEventListener('click', () => mountDataMaturityDashboard());
        host.querySelector('#ndm-btn-report')?.addEventListener('click', () => {
          global.open(`${API_BASE}/api/data-maturity/report.html`, '_blank', 'noopener');
        });
        return payload;
      })
      .catch(() => {
        body.innerHTML = '<p class="esr-muted">Maturité des données nationales indisponible.</p>';
        return null;
      });
  }

  global.DataMaturityDashboard = { mountDataMaturityDashboard };
  // Compat ESR : aussi exposé via TerritorialImpactUI si déjà présent
  global.TerritorialImpactUI = global.TerritorialImpactUI || {};
  global.TerritorialImpactUI.mountDataMaturityDashboard = mountDataMaturityDashboard;
})(window);
